### PR TITLE
🧾 Feat/1160 Treatment Detail records added to Exception Manifest

### PR DESCRIPTION
**Pull request by demariadaniel**
## Link to Issue
Unit test & TSV output in ticket
[#1160](https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/gh/icgc-argo/argo-clinical/1160)
## Description
- Adds Treatment Detail Exception Records to Manifest
- Moves main exception manifest code to own file
- Updates Sorting for Missing Entity & Treatment Detail exception records
- Updates Unit Tests

### Type of Change

- [ ] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

### Checklist before requesting review:

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/exception/exception-manifest/index.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+} from './types';
+
+import {
+	createProgramExceptions,
+	mapEntityExceptionRecords,
+	sortExceptionRecordsBySubmitterId,
+	sortExceptionRecordsByEntityId,
+} from './util';
+
+export const createExceptionManifest = async (
```
Comment (demariadaniel) [inline src/exception/exception-manifest/index.ts:43]: Moved majority of function out of `/submission` into `/exception`

File: src/exception/exception-manifest/util.ts [hunk 2]
```diff
@@ -157,15 +159,26 @@ export const mapEntityExceptionRecords = (programId: string, donors: DeepReadonl
 	return entityRecord;
 };
 
-export const sortExceptionRecordsBySubmitterId = (
+function baseExceptionSort<Exception>(
```
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:162]: why the change to use the `function` keyword?

Comment (demariadaniel) [reply src/exception/exception-manifest/util.ts:162]: Oh haha this was just following TS Docs example
It could be changed to:

`export const baseExceptionSort = <Exception>(
	first: Exception,
	next: Exception,
	key: keyof Exception,
): number => {
	return first[key] === next[key] ? 0 : first[key] > next[key] ? 1 : -1;
};`

File: src/submission/exceptions/exceptions.ts [hunk 3]
```diff
@@ -308,54 +298,20 @@ export async function getExceptionManifestRecords(
...
 	);
 
-	// Exceptions only store submitterIds, so all submitterIds have to be collected before we can filter exceptions
-	const submitterDonorIds = donors.map((donor) => donor.submitterId);
+	const { programException, entityException } = await queryForExceptions(programId);
 
-	const { programException, entityException: entityPropertyException } = await queryForExceptions(
-		programId,
-	);
+	const programExceptions = programException?.exceptions || [];
```
Comment (ciaranschutte) [inline src/submission/exceptions/exceptions.ts:303]: great error checking! there's also error checking in the `createExceptionManifest` function.
could be good to do all the checking for ` {
		programExceptions,
		entityException,
		missingEntityException,
		treatmentDetailException,
	});` first and then pass only what you need into the `create` function

Comment (ciaranschutte) [reply src/submission/exceptions/exceptions.ts:303]: (same kind of error checking, ie. check if you got the exception record, check if result is correct etc)

**Review by ciaranschutte — APPROVED** [review]
couple suggestions.lgtm.

File: src/exception/exception-manifest/index.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+import {
+	createProgramExceptions,
+	mapEntityExceptionRecords,
+	sortDonorExceptionRecordsBySubmitterId,
+	sortEntityExceptionRecordsBySubmitterId,
+	sortExceptionRecordsByEntityId,
+} from './util';
+
+export const createExceptionManifest = async (
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/exception/exception-manifest/index.ts:44]: Removed; left over from an earlier state of the code

File: src/submission/exceptions/exceptions.ts (hunk 3 above)
Comment (demariadaniel) [reply src/submission/exceptions/exceptions.ts:303]: That was the original intent, to separate the two functions as you describe. Revisited and updated!

**Review by ciaranschutte — APPROVED** [review]
